task/mongo-optimization-task.js

```diff
@@ -39,267 +41,265 @@ async function before(db) {
-                        "$or" : [
-                            {
-                                "contacts.win_vendor.value" : {
-                                    "$in" : [
-                                        50
-                                    ]
-                                }
-                            },
-                            {
-                                "contacts.win_vendor.name" : "ADP"
-                            }
-                        ]
+            "from": "clientCriteria",
+            let: { v1: "$criteria_value" },
+            pipeline: [
+                {
+                    $match: {
+                        "versions.initiativeId": ObjectId("58af4da0b310d92314627290")
                     }
-                ]
-            }
-        },
-        {
-            "$lookup" : {
-                "from" : "clientCriteria",
-                "localField" : "criteria_value",
-                "foreignField" : "value",
-                "as" : "criteria"
-            }
-        },
-        {
-            "$unwind" : "$criteria"
-        },
-        {
-            "$unwind" : "$criteria.versions"
-        },
-        {
-            "$match" : {
-                "criteria.versions.initiativeId" : ObjectId("58af4da0b310d92314627290")
-            }
-        },
-        {
-            "$group" : {
-                "_id" : "$contacts.questions.answers.primary_answer_value",
-                "answer_value" : {
-                    "$first" : "$contacts.questions.answers.primary_answer_value"
                 },
-                "answer_text" : {
-                    "$first" : "$contacts.questions.answers.primary_answer_text"
+                {
+                    $project: {
+                        "value": 1,
+                        "label": 1,
+                        "definition": 1,
+                        "versions.definition": 1,
+                    }
                 },
-                "answers" : {
-                    "$push" : {
-                        "c" : "$contacts.id",
-                        "question_category" : "$contacts.questions.category_id",
-                        "question_id" : "$contacts.questions.id",
-                        "ins" : "$contacts.questions.answers.loopInstances.loop_instance",
-                        "answer_value" : "$contacts.questions.answers.primary_answer_value",
-                        "selected" : "$contacts.questions.answers.loopInstances.is_selected",
-                        "value" : "$criteria_value",
-                        "text" : "$criteria.label",
-                        "definition" : {
-                            "$ifNull" : [
-                                "$criteria.versions.definition",
-                                "$criteria.definition"
-                            ]
-                        }
+                {
+                    $match: {
+                        $expr: { $eq: ["$value", "$$v1"] }
```

Comment:
can you add the query in that lookup in the first `$match` ?

Reply:
I tried, but it works slower.

Reply:
1. That can be ~ the same... but it's weird if it's slower. Also order of fields in index is important. 
`await db.collection('clientCriteria').createIndex({'value': 1,'versions.initiativeId': 1});`
Try vise versa. 

2. limit at the end of pipeline can help in some cases when you need only one item in result of `lookup`